tools/UpdateModules.ps1

```diff
@@ -65,12 +66,80 @@ function Create-ModulePsm1
      $template = $template -replace "%MODULE-NAME%", $file.BaseName
      $template = $template -replace "%DATE%", [string](Get-Date)
      $template = $template -replace "%IMPORTED-DEPENDENCIES%", $importedModules
+
+     $resourceGroupCompleterCommands = Find-CompleterAttribute -CompleterName "ResourceGroupCompleterAttribute" -ModuleMetadata $ModuleMetadata -ModulePath $ModulePath -isRMModule $isRMModule
+     $template = $template -replace "%RGCCOMMANDS%", $resourceGroupCompleterCommands
+
+     $locationCompleterCommands = Find-CompleterAttribute -CompleterName "LocationCompleterAttribute" -ModuleMetadata $ModuleMetadata -ModulePath $ModulePath -isRMModule $isRMModule
+     $template = $template -replace "%LCCOMMANDS%", $locationCompleterCommands
+
      Write-Host "Writing psm1 manifest to $templateOutputPath"
      $template | Out-File -FilePath $templateOutputPath -Force
      $file = Get-Item -Path $templateOutputPath
   }
 }
 
+function Find-CompleterAttribute
+{
+    [CmdletBinding()]
+    param(
+        [string]$CompleterName,
+        [Hashtable]$ModuleMetadata,
+        [string]$ModulePath,
+        [bool]$isRMModule
+    )
+    PROCESS
+    {
+        if ($isRMModule)
+        {
+            $nestedModules = $ModuleMetadata.NestedModules
+            $AllCmdlets = @()
+            $nestedModules | ForEach-Object {
+                $dllPath = Join-Path -Path $ModulePath -ChildPath $_
+                $Assembly = [Reflection.Assembly]::LoadFrom($dllPath)
+                $dllCmdlets = $Assembly.GetTypes() | Where-Object {$_.CustomAttributes.AttributeType.Name -contains "CmdletAttribute"}
+                $AllCmdlets += $dllCmdlets
+            }
+
+            $constructedCommands = "@("
+            $AllCmdlets | ForEach-Object {
+                $currentCmdlet = $_
+                $parameters = $_.GetProperties()
+                $parameters | ForEach-Object {
+                    $completerAttribute = $_.CustomAttributes | Where-Object {$_.AttributeType.Name -eq $CompleterName}
+                    if ($completerAttribute -ne $null) {
+                        $constructedCommands += "@('" + $currentCmdlet.GetCustomAttributes("System.Management.Automation.CmdletAttribute").VerbName + "-" + $currentCmdlet.GetCustomAttributes("System.Management.Automation.CmdletAttribute").NounName + "', "
+                        $constructedCommands += "'" + $_.Name + "', "
+                        if ($completerAttribute.ConstructorArguments.Count -eq 0) 
```

Comment:
this would be simplified if you store the cmdletattribute in a varaible:

```
$AtrributeTypeName="System.Management.Automation.CmdletAttribute"
$cmdletAttribute = $currentCmdlet.GetCustomAttributes($attributeTypeName)
```

It also seems like the result should be a list of hashtables ```@{CmdletName="<name>"; CmdletType="<typeName>"}```  this will make any code that uses this list a bit easier to read.

Finally, I think we should add some methods to the attribute itself to generate the list of constructor argument values, i.e. $completerAttribute.GetConstructorArgumentList() .

Reply:
I wrote the code adding a method for generating the list of constructor arguments, but it actually ended up being more convoluted because I had to create a new LocationCompleterAttribute object and convert the list of strings from c# to powershell.  I'll leave this how it is for now, but let me know if you would rather creating a new method.